### PR TITLE
first PR

### PR DESCRIPTION
**Pull request by chrislewispac**
this is the first PR of this small library which is intended to be a side project of mine to be used in other side projects where I want a simple, efficient, bug-free router for small SPA projects where I am not using any frameworks (no react/vue/angular). It is an exercise in improving my coding, publishing my first npm module, and having something I can use in the future. 

I appreciate your feedback. <3

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+node_modules
+coverage
```
[the remark this answers is not in the document]
Comment (chrislewispac) [reply .gitignore]: thank you. 

fixed in 8792073add02e88a36908e0d24b4e4458f503916

File: cypress.json [hunk 2]
```diff
@@ -0,0 +1 @@
+{}
```
[the remark this answers is not in the document]
Comment (chrislewispac) [reply cypress.json]: thank you.

fixed in 8792073add02e88a36908e0d24b4e4458f503916

File: dist/main.esm.js [hunk 3]
```diff
@@ -0,0 +1,2 @@
...
+//# sourceMappingURL=main.esm.js.map
```
[the remark this answers is not in the document]
Comment (chrislewispac) [reply dist/main.esm.js:2]: thank you. fixed in 8792073add02e88a36908e0d24b4e4458f503916

File: examples/webpack.config.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        rules: [
+        {
+            test: /\.css$/,
+            use: [
+            'style-loader',
+            'css-loader',
+            ],
+        },
+        {
+            test: /\.(png|svg|jpg|jpeg|gif)$/,
```
[the remark this answers is not in the document]
Comment (chrislewispac) [reply examples/webpack.config.js]: thank you, fixed in 8792073add02e88a36908e0d24b4e4458f503916

File: jest.config.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
+module.exports = {
+    "roots": [
+      "<rootDir>/src"
+    ],
+    "testMatch": [
+      "**/__tests__/**/*.+(ts|tsx|js)",
```
[the remark this answers is not in the document]
Comment (chrislewispac) [reply jest.config.js]: thank you, fixed in 8792073add02e88a36908e0d24b4e4458f503916

File: package.json [hunk 6]
```diff
@@ -47,7 +47,7 @@
   },
```
[the remark this answers is not in the document]
Comment (chrislewispac) [reply package.json:46]: oops. I had it on npm test but not watch-test. fixed/fixing. thank you

File: package.json [hunk 7]
```diff
@@ -47,7 +47,7 @@
...
   "jest": {
     "globals": {
-      "PKG_VERSION":"$npm_package_version"
+      "PKG_VERSION": "$npm_package_version"
```
[the remark this answers is not in the document]
Comment (chrislewispac) [reply package.json:49]: thank you. 👍

File: src/main.test.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
+import {Router} from './main'
```
[the remark this answers is not in the document]
Comment (chrislewispac) [reply src/main.test.ts:1]: fixed the xo issue

File: src/main.test.ts [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+beforeEach(async () => {
+    r = new Router({rootNode: root, baseRoute: '/', baseHandler})
+})
+
+describe('Router', () => {
+    it('Router has a method which providers the version', () => {
+        expect(r.version).toBe(version)
+    })
+    it('Router will initialize with a trie with only a root node', () => {
+        expect(r.trie.root.path).toBe("/")
```
[the remark this answers is not in the document]
Comment (chrislewispac) [reply src/main.test.ts:18]: you're right. thank you

File: src/trie.ts [hunk 10]
```diff
@@ -0,0 +1,102 @@
+const defaultRootHandler = () => { return "<h1>Thank you for using this router, please declare a baseHandler function during initialization.</h1>" }
+
+export class Trie {
+    private _root: TrieNode
+    private _defaultNode: TrieNode
+
+    constructor(config?: { baseRoute?: string, baseHandler?: Function }) {
+        const { baseRoute = "/", baseHandler = defaultRootHandler } = config || {}
+        this._root = new TrieNode({path: baseRoute, handler: baseHandler})
+        this._defaultNode = new TrieNode({path: '404', handler: () => {return "404 Node"}})
```
[the remark this answers is not in the document]
Comment (chrislewispac) [reply src/trie.ts:10]: this is a very good callout. I'm going to think about how I could move the trie implementation into its own stand-alone module without these cross-dependencies.

File: src/main.test.ts [hunk 11]
```diff
@@ -0,0 +1,58 @@
+import {Router} from './main'
+import {version} from './version'
+document.body.innerHTML = `<div id="root"></>`;
+
+const baseHandler = () => {return "Root Node"}
+const root = document.getElementById("root") 
+let r : Router;
+
+beforeEach(async () => {
```
[the remark this answers is not in the document]
Comment (chrislewispac) [reply src/main.test.ts:12]: thank you!

File: src/main.ts [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+	private readonly _trie: Trie;
+	private _root: Element;
+	private readonly _window: Window;
+
+	constructor({rootNode, baseHandler, baseRoute = '/'}: Config) {
+		// Need validators for these initial private variables
+		this._root = rootNode;
+		this._version = version;
+		this._trie = new Trie({baseRoute, baseHandler});
+		this._window = window;
```
[the remark this answers is not in the document]
Comment (chrislewispac) [reply src/main.ts:17]: agreed. I did this initially for testing (i didnt know how to have a browser env in my testing env).  #4

File: src/main.ts [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+		this.checkErrors();
+	}
+
+	addRoute(path: string, getPageContents: RouterHandler) {
+		const newRouteNode = this._trie.insert(path, getPageContents);
+
+		// NOT READY: handle the case of the initial route this seems like a bad way to do this
+		// but when you consider that adding a route is essentially initialization this might make sense
+		this.onRoute(window.location.pathname);
```
[the remark this answers is not in the document]
Comment (chrislewispac) [reply src/main.ts:26]: agreed #5 . Currently pulling trie into it's own npm module then will come back to these. THANK YOU for all of your help

File: src/main.ts [hunk 14]
```diff
@@ -0,0 +1,55 @@
+import {version} from './version';
```
[the remark this answers is not in the document]
Comment (danMateer) [reply src/main.ts:1]: > _Possible Blocking Issue (PR_BLCKNG)_

[the remark this answers is not in the document]
Comment (danMateer) [reply src/main.ts:1]: > _Suggestion (PR_SGGTN)_

**Pull request merged**